pyensign/exceptions.py

```diff
@@ -1,3 +1,93 @@
+import inspect
+from grpc import RpcError
+from functools import wraps
+
+##########################################################################
+## Decorators
+##########################################################################
+
+_except = (RpcError, AttributeError)
+
+
+def wrap_rpc(fn):
+    @wraps(fn)
+    def wrap(*args, **kwargs):
+        try:
+            return fn(*args, **kwargs)
+        except _except as e:
+            handle_client_error(e)
+
+    return wrap
+
+
+def wrap_async_rpc(coro):
+    @wraps(coro)
+    async def wrap(*args, **kwargs):
+        try:
+            return await coro(*args, **kwargs)
+        except _except as e:
+            handle_client_error(e)
+
+    return wrap
+
+
+def wrap_generator(fn):
+    @wraps(fn)
+    def wrap(*args, **kwargs):
+        try:
+            for rep in fn(*args, **kwargs):
+                yield rep
+        except _except as e:
+            handle_client_error(e)
+
+    return wrap
+
+
+def wrap_async_generator(coro):
+    @wraps(coro)
+    async def wrap(*args, **kwargs):
+        try:
+            async for rep in coro(*args, **kwargs):
+                yield rep
+        except _except as e:
+            handle_client_error(e)
+
+    return wrap
+
+
+def catch_rpc_error(coro):
+    """
+    Decorate coroutines to catch and handle gRPC client errors.
+    """
+
+    if inspect.isgeneratorfunction(coro):
+        return wrap_generator(coro)
+    elif inspect.isasyncgenfunction(coro):
+        return wrap_async_generator(coro)
+    elif inspect.iscoroutinefunction(coro):
+        return wrap_async_rpc(coro)
+    else:
+        return wrap_rpc(coro)
+
+
+##########################################################################
+## Error Handling
+##########################################################################
+
+
+def handle_client_error(e):
+    """
+    Provides error handling for errors raised by the gRPC client.
+    """
+
+    if isinstance(e, RpcError):
+        raise EnsignRPCError(e)
```

Comment:
Could you remind me what `raise EnsignRPCError(e) as e` does in Python again? E.g. do we want to preserve the stack from where the exception was raised somehow? I believe what happens here is that the stack will start at this `raise` statement.

Reply:
Looks like `raise EnsignRPCError(e)` is still chaining the errors and preserving the stack trace, but `raise EnsignRPCError("err msg") from e` might be the way we want to go so we can add more context in the future with the RPC error codes in place of the "err msg"

Reply:
Cheers!